service/matching/matchingEngine.go

```diff
@@ -444,10 +446,16 @@ func (e *matchingEngineImpl) DispatchSpooledTask(
 	unversionedOrigTaskQueue := newTaskQueueIDWithVersionSet(origTaskQueue, "")
 	// Redirect and re-resolve if we're blocked in matcher and user data changes.
 	for {
-		taskQueue, userDataChanged, err := e.redirectToVersionedQueueForAdd(
+		taskQueue, _, userDataChanged, err := e.redirectToVersionedQueueForAdd(
 			ctx, unversionedOrigTaskQueue, directive, stickyInfo)
 		if err != nil {
-			return err
+			if errors.Is(err, errUserDataDisabled) && directive.GetBuildId() == "" {
+				// Only fail tasks with compatiblity constraints when user data is disabled.
```

Comment:
could you reword this? "failing a task" is something kind of specific, that's not what we're doing here. we're just punting it back to the taskReader loop to be retried a second later